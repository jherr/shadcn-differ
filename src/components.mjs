import path from "node:path";
import fs from "node:fs";

const WHITELISTED_COMPONENTS = [
  "accordion",
  "alert",
  "alert-dialog",
  "aspect-ratio",
  "avatar",
  "badge",
  "breadcrumb",
  "button",
  "calendar",
  "card",
  "carousel",
  "chart",
  "checkbox",
  "collapsible",
  "command",
  "context-menu",
  "table",
  "dialog",
  "drawer",
  "dropdown-menu",
  "form",
  "hover-card",
  "input",
  "input-otp",
  "label",
  "menubar",
  "navigation-menu",
  "pagination",
  "popover",
  "progress",
  "radio-group",
  "resizable",
  "scroll-area",
  "select",
  "separator",
  "sheet",
  "skeleton",
  "slider",
  "sonner",
  "switch",
  "tabs",
  "textarea",
  "toast",
  "toggle",
  "toggle-group",
  "tooltip",
];

export function findComponentFiles(config, originalFiles) {
  const registryDependencies = [];
  const compDir = config.ui.replace("@/", config.isSrcDir ? "src/" : "");
  for (const { path: filePath } of originalFiles) {
    if (filePath.startsWith(compDir)) {
      const fileExtension = path.extname(filePath);
      const fileName = path.basename(filePath, fileExtension);
      if (
        (fileExtension === ".tsx" || fileExtension === ".jsx") &&
        WHITELISTED_COMPONENTS.includes(fileName)
      ) {
        registryDependencies.push(path.basename(filePath, fileExtension));
      }
    }
  }
  return registryDependencies;
}

export function readComponentsManifest(dir) {
  const manifestPath = path.join(dir, "./components.json");
  if (fs.existsSync(manifestPath)) {
    const config = JSON.parse(fs.readFileSync(manifestPath, "utf-8")).aliases;
    return config;
  } else {
    console.error("Components manifest not found");
    process.exit(1);
  }
}

export function getAliasedPaths(config) {
  return [
    config.components.replace("@/", ""),
    config.utils.replace("@/", ""),
    config.ui.replace("@/", ""),
    config.lib.replace("@/", ""),
    config.hooks.replace("@/", ""),
  ];
}

export function isBuiltinComponent(config, filePath) {
  if (filePath.startsWith(config.ui.replace("@/", ""))) {
    const component = path.basename(filePath, path.extname(filePath));
    return WHITELISTED_COMPONENTS.includes(component);
  }
  return false;
}
