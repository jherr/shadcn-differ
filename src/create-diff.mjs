import {
  findComponentFiles,
  getAliasedPaths,
  isBuiltinComponent,
} from "./components.mjs";
import { parseFilePath } from "./parse-file-path.mjs";

function addFile(output, config, inSrcDir, relativeFilePath, content) {
  if (!isBuiltinComponent(config, relativeFilePath)) {
    output.files.push(
      parseFilePath(inSrcDir, config, `./${relativeFilePath}`, content)
    );
  }
}

function addDependencies(
  output,
  initialPackageContents,
  currentPackageContents
) {
  const initialPackageJson = JSON.parse(initialPackageContents);
  const currentPackageJson = JSON.parse(currentPackageContents);

  output.dependencies = Object.keys(
    currentPackageJson.dependencies || {}
  ).filter((dep) => !initialPackageJson.dependencies.hasOwnProperty(dep));
  output.devDependencies = Object.keys(
    currentPackageJson.devDependencies || {}
  ).filter((dep) => !initialPackageJson.devDependencies.hasOwnProperty(dep));
}

function scanWithSrcDir(output, config, alteredFiles) {
  for (const { path, content } of alteredFiles) {
    if (path.startsWith("src/")) {
      addFile(output, config, true, path.replace("src/", ""), content);
    } else {
      addFile(output, config, false, path, content);
    }
  }
}

function isInAppDir(path) {
  return path.startsWith("app/");
}

function scanWithoutSrcDir(output, config, alteredFiles) {
  const aliasedPaths = getAliasedPaths(config);

  for (const { path, content } of alteredFiles) {
    addFile(
      output,
      config,
      aliasedPaths.includes(path) || isInAppDir(path),
      path,
      content
    );
  }
}

export function createDiff({
  name,
  config,
  alteredFiles,
  specificFiles,
  currentFiles,
  currentPackageJson,
}) {
  const output = {
    name,
    type: "registry:block",
    dependencies: [],
    devDependencies: [],
    registryDependencies: [],
    files: [],
    tailwind: {},
    cssVars: {},
    meta: {},
  };

  if (config.isSrcDir) {
    scanWithSrcDir(output, config, alteredFiles);
  } else {
    scanWithoutSrcDir(output, config, alteredFiles);
  }

  output.registryDependencies = findComponentFiles(config, currentFiles);

  addDependencies(output, specificFiles["./package.json"], currentPackageJson);

  return output;
}
