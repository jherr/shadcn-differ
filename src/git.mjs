import fs from "node:fs";
import path from "node:path";
import { execSync } from "node:child_process";
import ignore from "ignore";
import { start } from "node:repl";

const INITIAL_DIR = "_initial";

const EXCLUDE_DIRS = [
  "node_modules",
  "dist",
  "fonts",
  "build",
  "public",
  "static",
  ".next",
  ".git",
  INITIAL_DIR,
];

const EXCLUDE_FILES = [
  ".DS_Store",
  "next-env.d.ts",
  "package-lock.json",
  "yarn.lock",
  "pnpm-lock.yaml",
  "bun.lockb",
  "package.json",
  "tailwind.config.ts",
  "tailwind.config.js",
  "components.json",
  "favicon.ico",
];

function cloneInitialCommit() {
  deleteInitialDir();

  try {
    // Get the initial commit hash
    const initialCommit = execSync("git rev-list --max-parents=0 HEAD")
      .toString()
      .trim();

    // Clone the initial commit quietly
    execSync(`git worktree add -f ${INITIAL_DIR} ${initialCommit}`, {
      stdio: "ignore",
    });
  } catch (error) {
    console.error("Error cloning initial commit:", error.message);
    process.exit(1);
  }
}

function deleteInitialDir() {
  if (fs.existsSync(INITIAL_DIR)) {
    fs.rmSync(INITIAL_DIR, { recursive: true });

    try {
      execSync("git worktree prune", { stdio: "ignore" });
    } catch (error) {
      console.error("Error pruning git worktree:", error.message);
    }
  }
}

function checkIfFileIsChanged(relativeFilePath) {
  const initialFilePath = path.join(INITIAL_DIR, relativeFilePath);
  const fullPath = path.join(process.cwd(), relativeFilePath);
  if (!fs.existsSync(initialFilePath)) {
    return true; // New file
  }
  const currentContent = fs.readFileSync(fullPath, "utf-8");
  const initialContent = fs.readFileSync(initialFilePath, "utf-8");
  return currentContent !== initialContent;
}

export function scanForFiles(startDir, checkFile = false) {
  const foundFiles = [];

  let ignorer = () => false;
  if (fs.existsSync(path.join(startDir, ".gitignore"))) {
    const gitIgnore = ignore().add(
      fs.readFileSync(path.join(startDir, ".gitignore")).toString()
    );
    ignorer = (relativeFilePath) => {
      return gitIgnore.ignores(relativeFilePath);
    };
  }

  function scanDirectory(dir, relativePath = "") {
    const entries = fs.readdirSync(dir, { withFileTypes: true });

    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);
      const relativeFilePath = path.join(relativePath, entry.name);

      if (entry.isDirectory()) {
        if (!EXCLUDE_DIRS.includes(entry.name)) {
          scanDirectory(path.join(dir, entry.name), relativeFilePath);
        }
      } else if (
        !checkFile ||
        (checkFile && checkIfFileIsChanged(relativeFilePath))
      ) {
        if (!EXCLUDE_FILES.includes(entry.name) && !ignorer(relativeFilePath)) {
          foundFiles.push({
            path: relativeFilePath,
            content: fs.readFileSync(fullPath, "utf-8"),
          });
        }
      }
    }
  }

  scanDirectory(startDir);

  return foundFiles;
}

export function scanForAlteredFiles(specificFilesToReturn = []) {
  cloneInitialCommit();

  const alteredFiles = scanForFiles(process.cwd(), true);

  const specificFiles = specificFilesToReturn.reduce((out, file) => {
    const fullPath = path.join(process.cwd(), INITIAL_DIR, file);
    out[file] = fs.readFileSync(fullPath, "utf-8");
    return out;
  }, {});

  deleteInitialDir();

  return {
    alteredFiles,
    specificFiles,
  };
}

export function hasSrcDir(dir) {
  return fs.existsSync(path.join(dir, "src"));
}
