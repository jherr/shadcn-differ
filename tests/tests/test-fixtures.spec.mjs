import { describe, expect, test } from "vitest";
import fs from "node:fs";
import path from "node:path";

import { scanForFiles, hasSrcDir } from "../../src/git.mjs";
import { readComponentsManifest } from "../../src/components.mjs";
import { createDiff } from "../../src/create-diff.mjs";

const EXPECTED_FILES = [
  {
    path: "./.env.example",
    content: "example env;",
    type: "registry:example",
    target: "~/.env.example",
  },
  {
    path: "./non-src-sub/test.ts",
    content: "non - src - sub;\n",
    type: "registry:example",
    target: "~/non-src-sub/test.ts",
  },
  {
    path: "./app/page.tsx",
    content: "page;\n",
    type: "registry:example",
    target: "./app/page.tsx",
  },
  {
    path: "./components/comp.tsx",
    content: "comp;\n",
    type: "registry:block",
    target: undefined,
  },
  {
    path: "./components/ui/new-ui-comp.tsx",
    content: "new-ui-comp;",
    type: "registry:ui",
    target: undefined,
  },
  {
    path: "./hooks/hook.ts",
    content: "hook;\n",
    type: "registry:hook",
    target: undefined,
  },
  {
    path: "./lib/lib.ts",
    content: "lib;\n",
    type: "registry:lib",
    target: undefined,
  },
];

function createDiffRequest(dir) {
  const currentFiles = scanForFiles(dir);

  const config = readComponentsManifest(dir);
  config.isSrcDir = hasSrcDir(dir);

  const specificFiles = {
    "./package.json": fs.readFileSync(
      path.join(dir, "./package.json"),
      "utf-8"
    ),
  };

  return {
    name: "example",
    config,
    alteredFiles: currentFiles,
    specificFiles,
    currentFiles,
    currentPackageJson: specificFiles["./package.json"],
  };
}

describe("fixture tests", () => {
  test("should diff a src directory project", () => {
    const diff = createDiff(createDiffRequest("./tests/fixtures/src-dir"));
    expect(diff).toMatchSnapshot();
  });
  test("should diff a non-src directory project", () => {
    const diff = createDiff(createDiffRequest("./tests/fixtures/non-src-dir"));
    expect(diff).toMatchSnapshot();
  });
});
