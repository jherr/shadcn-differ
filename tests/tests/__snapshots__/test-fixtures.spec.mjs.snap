// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fixture tests > should diff a non-src directory project 1`] = `
{
  "cssVars": {},
  "dependencies": [],
  "devDependencies": [],
  "files": [
    {
      "content": "example env;",
      "path": "./.env.example",
      "target": "~/.env.example",
      "type": "registry:example",
    },
    {
      "content": "page;
",
      "path": "./app/page.tsx",
      "target": "./app/page.tsx",
      "type": "registry:example",
    },
    {
      "content": "comp;
",
      "path": "./components/comp.tsx",
      "target": undefined,
      "type": "registry:block",
    },
    {
      "content": "new-ui-comp;",
      "path": "./components/ui/new-ui-comp.tsx",
      "target": undefined,
      "type": "registry:ui",
    },
    {
      "content": "hook;
",
      "path": "./hooks/hook.ts",
      "target": undefined,
      "type": "registry:hook",
    },
    {
      "content": "lib;
",
      "path": "./lib/lib.ts",
      "target": undefined,
      "type": "registry:lib",
    },
    {
      "content": "non - src - sub;
",
      "path": "./non-src-sub/test.ts",
      "target": "~/non-src-sub/test.ts",
      "type": "registry:example",
    },
  ],
  "meta": {},
  "name": "example",
  "registryDependencies": [
    "select",
  ],
  "tailwind": {},
  "type": "registry:block",
}
`;

exports[`fixture tests > should diff a src directory project 1`] = `
{
  "cssVars": {},
  "dependencies": [],
  "devDependencies": [],
  "files": [
    {
      "content": "example env;",
      "path": "./.env.example",
      "target": "~/.env.example",
      "type": "registry:example",
    },
    {
      "content": "non - src - sub;
",
      "path": "./non-src-sub/test.ts",
      "target": "~/non-src-sub/test.ts",
      "type": "registry:example",
    },
    {
      "content": "page;
",
      "path": "./app/page.tsx",
      "target": "./app/page.tsx",
      "type": "registry:example",
    },
    {
      "content": "comp;
",
      "path": "./components/comp.tsx",
      "target": undefined,
      "type": "registry:block",
    },
    {
      "content": "new-ui-comp;",
      "path": "./components/ui/new-ui-comp.tsx",
      "target": undefined,
      "type": "registry:ui",
    },
    {
      "content": "hook;
",
      "path": "./hooks/hook.ts",
      "target": undefined,
      "type": "registry:hook",
    },
    {
      "content": "lib;
",
      "path": "./lib/lib.ts",
      "target": undefined,
      "type": "registry:lib",
    },
  ],
  "meta": {},
  "name": "example",
  "registryDependencies": [
    "select",
  ],
  "tailwind": {},
  "type": "registry:block",
}
`;
